airbyte-cdk/python/airbyte_cdk/sources/streams/core.py

```diff
@@ -160,19 +147,31 @@ def read_incremental(  # type: ignore  # ignoring typing for ConnectorStateManag
                 ):
                     record_data = record_data_or_message if isinstance(record_data_or_message, Mapping) else record_data_or_message.record
                     stream_state = self.get_updated_state(stream_state, record_data)
-                    checkpoint_interval = self.state_checkpoint_interval
                     record_counter += 1
-                    if checkpoint_interval and record_counter % checkpoint_interval == 0:
-                        yield self._checkpoint_state(stream_state, state_manager, per_stream_state_enabled)
+
+                    if sync_mode == SyncMode.incremental:
+                        # Checkpoint intervals are a bit controversial, but see below comment about why we're gating it right now
+                        checkpoint_interval = self.state_checkpoint_interval
+                        if checkpoint_interval and record_counter % checkpoint_interval == 0:
+                            yield self._checkpoint_state(stream_state, state_manager)
 
                     if internal_config.is_limit_reached(record_counter):
                         break
 
-            yield self._checkpoint_state(stream_state, state_manager, per_stream_state_enabled)
+            if sync_mode == SyncMode.incremental:
+                # Even though right now, only incremental streams running as incremental mode will emit periodic checkpoints. Rather than
+                # overhaul how refresh interacts with the platform, this positions the code so that once we want to start emitting
+                # periodic checkpoints in full refresh mode it can be done here
+                yield self._checkpoint_state(stream_state, state_manager)
+
+        if not has_slices or sync_mode == SyncMode.full_refresh:
+            if sync_mode == SyncMode.full_refresh:
+                # We use a dummy state if there is no suitable value provided by full_refresh streams that do not have a valid cursor.
+                # Incremental streams running full_refresh mode emit a meaningful state
+                stream_state = stream_state or {"sync_mode": "full_refresh"}
```

Comment:
After testing on the platform, the platform will persist stream state regardless of the sync mode it's running in. It might actually be better to store `{}` rather than a bogus value that isn't usable.

The other alternative is a best attempt at a cursor using the `emitted_at` of the last record. My reservation here is that gives a false perception of an accurate cursor value which is misleading for all our full refresh streams